keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,201 @@
+---
+title: Union types
+authors:
+  - "@apelisse"
+owning-sig: sig-api-machinery
+participating-sigs:
+reviewers:
+approvers:
+- @lavalamp
+editor: TBD
+creation-date: 2019-03-25
+last-updated: 2019-03-25
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+replaces:
+superseded-by:
+---
+
+# Union types
+
+## Summary
+
+Modern data model definitions like OpenAPI v3 and protobuf (versions 2 and 3)
+have a keyword to implement “oneof” or "union". They allow APIs to have a better
+semantic, typically as a way to say “only one of the given fields can be
+set”. We currently have multiple occurrences of this semantic in kubernetes core
+types, at least:
+- DeploymentStrategy is a structure that has a discrminator
```

Comment:
nit: put this example second--don't include a discriminator in the first example, discriminator with only one choice is perhaps a surprising thing to include for some people.